File: cpp/0200_number_of_islands.cpp
// DFS solution
class Solution {
public:
    int numIslands(vector<vector<char>>& grid) {
      int y = grid.size();
      if (y == 0)
        return 0;
      int x = grid[0].size();
      

      int res = 0;

      for(int i = 0; i < y ; i++) {
        for (int j = 0; j < x; j++) {
          if (grid[i][j] == '1') {
            ++res;
            dfs(i, j, grid);
          }
        }
      }
      return res;
    }
    void dfs(int r, int c, vector<vector<char>>& grid) {
      grid[r][c] = '0';
    
      if (r - 1 >= 0 && grid[r-1][c] == '1') dfs(r - 1, c, grid);
      if (r + 1 < grid.size() && grid[r+1][c] == '1') dfs(r + 1, c, grid);
      if (c - 1 >= 0 && grid[r][c-1] == '1') dfs(r, c - 1, grid);
      if (c + 1 < grid[0].size() && grid[r][c+1] == '1') dfs(r, c + 1, grid);
    }
};

// BFS solution
class Solution {
public:
    int numIslands(vector<vector<char>>& grid) {
      int y = grid.size();
      if (y == 0)
        return 0;
      int x = grid[0].size();
    
      int res = 0;
      for (int i = 0; i < y; i++) {
        for (int j = 0; j < x; j++) {
          if(grid[i][j] == '1') {
            ++res;
            grid[i][j] == '0';
            queue<pair<int, int>> next;
            next.push({i, j});
            while(!next.empty()) {
              auto top = next.front();
              next.pop();
              
              if (top.first - 1 >= 0 && grid[top.first-1][top.second] == '1') {
                next.push({top.first-1, top.second});
                grid[top.first-1][top.second] = '0';
              }
              if (top.first + 1 < grid.size() && grid[top.first+1][top.second] == '1') {
                next.push({top.first+1, top.second});
                grid[top.first+1][top.second] = '0';
              }
              if (top.second - 1 >= 0 && grid[top.first][top.second-1] == '1') {
                next.push({top.first, top.second-1});
                grid[top.first][top.second-1] = '0';
              }
              if (top.second + 1 < grid[0].size() && grid[top.first][top.second+1] == '1') {
                next.push({top.first, top.second+1});
                grid[top.first][top.second+1] = '0';
              }
            }
          }
        }
      }

      return res;
    }
    
};

// BFS. use more space but little performance improvement.
class Solution {
public:
    int numIslands(vector<vector<char>>& grid) {
      int y = grid.size();
      if (y == 0)
        return 0;
      int x = grid[0].size();
    
      int res = 0;
      for (int i = 0; i < y; i++) {
        for (int j = 0; j < x; j++) {
          if(grid[i][j] == '1') {
            ++res;
            grid[i][j] = '0';
            // queue<pair<int, int>> next;
            queue<int> next_i;
            queue<int> next_j;
            // next.push({i, j});
            next_i.push(i);
            next_j.push(j);
            while(!next_i.empty()) {
              // top = next.front();
              // next.pop();
              int r = next_i.front();
              int c = next_j.front();
              next_i.pop();
              next_j.pop();
              
              if (r - 1 >= 0 && grid[r-1][c] == '1') {
                // next.push({top.first-1, top.second});
                next_i.push(r-1);
                next_j.push(c);
                grid[r-1][c] = '0';
              }
              if (r + 1 < grid.size() && grid[r+1][c] == '1') {
                // next.push({top.first+1, top.second});
                next_i.push(r+1);
                next_j.push(c);
                grid[r+1][c] = '0';
              }
              if (c - 1 >= 0 && grid[r][c-1] == '1') {
                // next.push({top.first, top.second-1});
                next_i.push(r);
                next_j.push(c-1);
                grid[r][c-1] = '0';
              }
              if (c + 1 < grid[0].size() && grid[r][c+1] == '1') {
                // next.push({top.first, top.second+1});
                next_i.push(r);
                next_j.push(c+1);
                grid[r][c+1] = '0';
              }
            }
          }
        }
      }

      return res;
    }
    
};

File: python/0200_number_of_islands.py
def numIslands(grid):
    n_row = len(grid)
    if n_row == 0:
        return 0
    n_col = len(grid[0])
    # f_grid = [i for item in grid for i in item]
    queue = []
    res = 0
    # for i in range(len(f_grid)):
    for i in range(n_row):
        for j in range(n_col):
            # if f_grid[i] == '1':
            if grid[i][j] == '1':
                res += 1
                grid[i][j] = '0'
                # f_grid[i] = '0'
                queue.append((i, j))
                # queue.append(i)
                while queue:
                    row, col = queue.pop(0)
                    # ind = queue.pop(0)
                    # left
                    if col > 0 and grid[row][col - 1] == '1':
                        queue.append((row, col - 1))
                        grid[row][col - 1] = '0'
                    # if ind % n_col != 0 and f_grid[ind - 1] == '1':
                    #     queue.append(ind - 1)
                    #     f_grid[ind - 1] = '0'
                    # right
                    if col + 1 < n_col and grid[row][col + 1] == '1':
                        queue.append((row, col + 1))
                        grid[row][col + 1] = '0'
                    # if (ind + 1) % n_col != 0 and f_grid[ind + 1] == '1':
                    #     queue.append(ind + 1)
                    #     f_grid[ind + 1] = '0'
                    # up
                    if row > 0 and grid[row - 1][col] == '1':
                        queue.append((row - 1, col))
                        grid[row - 1][col] = '0'
                    # if ind > n_col and f_grid[ind - n_col] == '1':
                    #     queue.append(ind - n_col)
                    #     f_grid[ind - n_col] = '0'
                    # down
                    if row + 1 < n_row and grid[row + 1][col] == '1':
                        queue.append((row + 1, col))
                        grid[row + 1][col] = '0'
                    # if ind + n_col < len(f_grid) and f_grid[ind + n_col] == '1':
                    #     queue.append(ind + n_col)
                    #     f_grid[ind + n_col] = '0'
    return res


print(numIslands([["1","1","1"],["0","1","0"],["1","1","1"]]))
print(numIslands([['1','1']]))
print(numIslands([
['1','1','0','0','0'],
['1','1','0','0','0'],
['0','0','1','0','0'],
['0','0','0','1','1']
]))
print(numIslands([]))
print(numIslands([['1','1','0','1','0']]))
print(numIslands([['1'],['1'],['0'],['1'],['0']]))
print(numIslands([["1","1","1","1","1","0","1","1","1","1","1","1","1","1","1","0","1","0","1","1"],
                  ["0","1","1","1","1","1","1","1","1","1","1","1","1","0","1","1","1","1","1","0"],
                  ["1","0","1","1","1","0","0","1","1","0","1","1","1","1","1","1","1","1","1","1"],
                  ["1","1","1","1","0","1","1","1","1","1","1","1","1","1","1","1","1","1","1","1"],
                  ["1","0","0","1","1","1","1","1","1","1","1","1","1","1","1","1","1","1","1","1"],
                  ["1","0","1","1","1","1","1","1","0","1","1","1","0","1","1","1","0","1","1","1"],
                  ["0","1","1","1","1","1","1","1","1","1","1","1","0","1","1","0","1","1","1","1"],
                  ["1","1","1","1","1","1","1","1","1","1","1","1","0","1","1","1","1","0","1","1"],
                  ["1","1","1","1","1","1","1","1","1","1","0","1","1","1","1","1","1","1","1","1"],
                  ["1","1","1","1","1","1","1","1","1","1","1","1","1","1","1","1","1","1","1","1"],
                  ["0","1","1","1","1","1","1","1","0","1","1","1","1","1","1","1","1","1","1","1"],
                  ["1","1","1","1","1","1","1","1","1","1","1","1","1","1","1","1","1","1","1","1"],
                  ["1","1","1","1","1","1","1","1","1","1","1","1","1","1","1","1","1","1","1","1"],
                  ["1","1","1","1","1","0","1","1","1","1","1","1","1","0","1","1","1","1","1","1"],
                  ["1","0","1","1","1","1","1","0","1","1","1","0","1","1","1","1","0","1","1","1"],
                  ["1","1","1","1","1","1","1","1","1","1","1","1","0","1","1","1","1","1","1","0"],
                  ["1","1","1","1","1","1","1","1","1","1","1","1","1","0","1","1","1","1","0","0"],
                  ["1","1","1","1","1","1","1","1","1","1","1","1","1","1","1","1","1","1","1","1"],
                  ["1","1","1","1","1","1","1","1","1","1","1","1","1","1","1","1","1","1","1","1"],
                  ["1","1","1","1","1","1","1","1","1","1","1","1","1","1","1","1","1","1","1","1"]]))
